rskj-core/src/main/java/co/rsk/rpc/ExecutionBlockRetriever.java

```diff
@@ -69,27 +66,26 @@ public BlockResult retrieveExecutionBlock(String bnOrId) {
         }
 
         if (PENDING_ID.equals(bnOrId)) {
-            Optional<Block> latestBlock = minerServer.getLatestBlock();
-            if (latestBlock.isPresent()) {
-                return newBlockResult(latestBlock.get());
+            Block bestBlock = blockchain.getBestBlock();
+            BlockResult result = cachedResult;
+            // optimistic check without the lock
+            if (result != null && result.getBlock().getParentHash().equals(bestBlock.getHash())) {
+                return result;
             }
 
-            Block bestBlock = blockchain.getBestBlock();
-            if (cachedBlock == null || !bestBlock.isParentOf(cachedBlock)) {
-
-                // If the miner server is not running there is no one to update the mining mainchain view,
-                // thus breaking eth_call with 'pending' parameter
-                //
-                // This is just a provisional fix not intended to remain in the long run
-                if (!minerServer.isRunning()) {
-                    miningMainchainView.addBest(bestBlock.getHeader());
+            synchronized (pendingBlockLock) {
```

Comment:
I think we also have some duplication here with the lines above comparing BB with cache, we could maybe extract a common method `getValidCachedResult` or similar.

Reply:
done